### ISSUE TITLE
100% system load on macOS Catalina (too eager incorrect case diagnostic)

### ISSUE THREAD

**Issue by neilyoung** [rust-analyzer/rust-analyzer#6433]
...caused by a process `rust-analyzer-mac`.

Any idea how to throttle that?

![image](https://user-images.githubusercontent.com/731020/97804104-f5649000-1c4d-11eb-9a86-d28e86d5952f.png)

**Comment by lnicola** [comment]
Does it happen on a specific project?

**Comment by lnicola** [comment]
You might be able to use Instruments (?) to see what it's doing, but most likely it's `chalk` going awry on some trait-heavy code. I think there are a couple of open issues on that theme.

**Comment by neilyoung** [comment]
It happens while I'm editing a rust file (of a bigger project) in VSCode. Yes, I can check with Instruments.  Any special wish what I should examine?

**Comment by lnicola** [comment]
No, I'm not familiar with the app.

**Comment by neilyoung** [comment]
Ok, I have now checked with activity monitor. The fan of the system is completely off, if VSCode and the rust script is not on top of the desk, but the fan immediately begins to roar and the load is up to 100% for rust-analyzer-mac, once I just hit "Enter" in the window. 

With every editing step I see a "cargo check" running in the lower status line. Making a snapshot of what happens over 2.5 secs finds rust-analyzer-mac having about 20 threads running, most of them hanging in "wait" for a mutex, but at least one is very active with sort of grammar analysis whereas most the time is spend in "ide_db::search::FindUsages:"

Does that make sense?

![image](https://user-images.githubusercontent.com/731020/97811763-dfba8f00-1c7c-11eb-826c-da13d3dbf14e.png)

**Comment by lnicola** [comment]
So it looks like an "incorrect case" assist is trying to compute a fix and needs to find all references to a name in order to do that.

@matklad I thought assists were resolved lazily, on request?

**Comment by matklad** [comment]
Huh, excellent catch! Those are diagnostics, and they are not resolved lazily. 

We probably should lazy-ifiy that, cc @popzxc

**Comment by neilyoung** [comment]
> Do you have auto-saving enabled?

Yes, this is enabled.

**Comment by popzxc** [comment]
Huh, I'll check what I can do.

**Comment by popzxc** [comment]
I think I identified the cause (thanks for the call graph!) and will send a fix today.

**Comment by neilyoung** [comment]
Looking forward :) Good luck

**Comment by neilyoung** [comment]
OK, fetched and compiled. I need a hint where to deploy the binaries. Using 

![image](https://user-images.githubusercontent.com/731020/98124262-3e694e00-1eb3-11eb-8549-3dc8b32c0572.png)


in VSCode. TIA

**Comment by lnicola** [comment]
@neilyoung you have two options:

1. compile it with `cargo xtask --install` and set `rust-analyzer.serverPath` to `~/.cargo/bin/rust-analyzer`
2. take the compiled binary and copy it to the extension directory (https://rust-analyzer.github.io/manual.html#vs-code, scroll down a little)

**Comment by popzxc** [comment]
Another option is to compile binary via `carbo build --release` and set `rust-analyzer.serverPath` to the `$FOLDER_WITH_RA/target/release/rust-analyzer`

**Comment by neilyoung** [comment]
> Another option is to compile binary via `carbo build --release` and set `rust-analyzer.serverPath` to the `$FOLDER_WITH_RA/target/release/rust-analyzer`

I was following this advice and nothing seemed to have changed, even though VSCode reacting on this change. However, a complete restart of VSCode with the new analyzer in the path definitely switched off my fans.

Thanks for the quick fix

**Comment by lnicola** [comment]
Yeah, you have to restart the language server (there's a command for it) or reload the window for it to take effect.

**Comment by neilyoung** [comment]
I did of course. In fact this was offered by VSC on detecting that I was tempering the config... Didn't help

**Comment by neilyoung** [comment]
Anyway, the fan is silent. You can close this issue

**Comment by popzxc** [comment]
Not until PR with the fix is merged, I believe :)

**Comment by neilyoung** [comment]
OK, sure. No problem

**Comment by neilyoung** [comment]
I noticed that the analyzer annotations are less obtrusive (what is great) now. Is this also something you did in your code?

**Comment by csnover** [comment]
Commenting mostly to add a cross-reference to issue #5344, since while they don’t seem to be related, until I actually reviewed my sampling data I wasn’t able to find this ticket due to the title using the term “system load” instead of “CPU”. GH issues doesn’t currently show this issue for searches like “high CPU” or “100% CPU”. Hopefully this helps others to find this ticket when they run into this problem.

A workaround is to add `incorrect-ident-case` to the `rust-analyzer.diagnostics.disabled` setting to disable this particular diagnostic.

**Comment by matklad** [comment]
Closing in favor of https://github.com/rust-analyzer/rust-analyzer/issues/8268
